### PR TITLE
Update AnalyzerRunner to support IIncrementalAnalyzer testing

### PR DESCRIPTION
**Pull request by sharwell**
This is a supporting change for upcoming profile-based optimization of SQLite and the incremental analyzer infrastructure for VS for Mac scenarios.

/cc @Therzok

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Would love to know what data this reveals once you collect it.  If you can make that information public, it would be very appreciated.  Thanks! :)

**Comment by sharwell** [comment]
> If you can make that information public ...

Literally the entire reason why this PR exists and the PRs from @Therzok are not yet merged is we are ensuring the public can reproduce the benchmarks that led to any changes. 😄

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+        private static async Task<AnalysisResult> GetProjectAnalysisResultAsync(
+            ImmutableArray<DiagnosticAnalyzer> analyzers,
+            Project project,
+            Options analyzerOptionsInternal,
+            CancellationToken cancellationToken)
+        {
+            WriteLine($"Running analyzers for {project.Name}", ConsoleColor.Gray);
+            if (analyzerOptionsInternal.RunConcurrent)
+            {
+                await Task.Yield();
```
Comment (Therzok) [inline src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: ConfigureAwait(false)?

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+            WriteLine($"Operation Block Actions:        {telemetry.OperationBlockActionsCount}", ConsoleColor.White);
+            WriteLine($"Operation Block Start Actions:  {telemetry.OperationBlockStartActionsCount}", ConsoleColor.White);
+            WriteLine($"Operation Block End Actions:    {telemetry.OperationBlockEndActionsCount}", ConsoleColor.White);
+
+            WriteLine($"Semantic Model Actions:         {telemetry.SemanticModelActionsCount}", ConsoleColor.White);
+            WriteLine($"Symbol Actions:                 {telemetry.SymbolActionsCount}", ConsoleColor.White);
+            WriteLine($"Symbol Start Actions:           {telemetry.SymbolStartActionsCount}", ConsoleColor.White);
+            WriteLine($"Symbol End Actions:             {telemetry.SymbolEndActionsCount}", ConsoleColor.White);
+            WriteLine($"Syntax Node Actions:            {telemetry.SyntaxNodeActionsCount}", ConsoleColor.White);
+            WriteLine($"Syntax Tree Actions:            {telemetry.SyntaxTreeActionsCount}", ConsoleColor.White);
```
Comment (Therzok) [inline src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: Maybe we can also add a metric for the number of diagnostics reported on each context - where applicable. (i.e. figure out busy/wasted work)

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        /// <see langword="false"/> to use the behavior configured for the specified <paramref name="project"/>.</param>
+        /// <param name="cancellationToken">The cancellation token that the task will observe.</param>
+        /// <returns>A list of analysis results inside the project</returns>
+        private static async Task<AnalysisResult> GetProjectAnalysisResultAsync(
+            ImmutableArray<DiagnosticAnalyzer> analyzers,
+            Project project,
+            Options analyzerOptionsInternal,
+            CancellationToken cancellationToken)
+        {
+            WriteLine($"Running analyzers for {project.Name}", ConsoleColor.Gray);
```
Comment (Therzok) [inline src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: Also log analyzer count?

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs (hunk 1 above)
Comment (sharwell) [reply src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: This is all just moved from Program.cs

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs (hunk 2 above)
Comment (sharwell) [reply src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: This is all just moved from Program.cs

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs (hunk 3 above)
Comment (sharwell) [reply src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: This is all just moved from Program.cs

File: src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            var processedProject = project.WithCompilationOptions(modifiedCompilationOptions);
+
+            try
+            {
+                Compilation compilation = await processedProject.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var newCompilation = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(compilation.SyntaxTrees);
+
+                var workspaceAnalyzerOptions = new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Options, project.Solution);
+                CompilationWithAnalyzers compilationWithAnalyzers = newCompilation.WithAnalyzers(analyzers, new CompilationWithAnalyzersOptions(workspaceAnalyzerOptions, null, analyzerOptionsInternal.RunConcurrent, logAnalyzerExecutionTime: true, reportSuppressedDiagnostics: analyzerOptionsInternal.ReportSuppressedDiagnostics));
+                var analystResult = await compilationWithAnalyzers.GetAnalysisResultAsync(cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Tools/AnalyzerRunner/DiagnosticAnalyzerRunner.cs]: This is all just moved from Program.cs

**Comment by sharwell** [comment]
@Therzok I have another issue to resolve with loading Roslyn.sln before this is ready to use. I'll let you know when I've updated the PR.

File: eng/targets/Settings.props [hunk 5]
```diff
@@ -30,7 +30,7 @@
     <UseRoslynAnalyzers Condition="'$(UseRoslynAnalyzers)' == ''">true</UseRoslynAnalyzers>
 
     <!-- Set to non-existent file to prevent common targets from importing Microsoft.CodeAnalysis.targets -->
-    <CodeAnalysisTargets>*none*</CodeAnalysisTargets>
+    <CodeAnalysisTargets>NON_EXISTENT_FILE</CodeAnalysisTargets>
```
Comment (sharwell) [inline eng/targets/Settings.props]: @dotnet/roslyn-infrastructure I meant to submit this as a separate pull request but I forgot to extract it. Please review this change.

File: src/Tools/AnalyzerRunner/AnalyzerRunner.csproj [hunk 6]
```diff
@@ -8,9 +8,35 @@
     <!-- Automatically generate the necessary assembly binding redirects -->
     <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
   </PropertyGroup>
+
+  <Choose>
```
Comment (sharwell) [inline src/Tools/AnalyzerRunner/AnalyzerRunner.csproj]: 📝 This entire block copied from ILTools.csproj. I'm not sure why it's required, but without it `<PlatformTarget>` ends up set to x86 and AnalyzerRunner runs as a 32-bit process.

File: src/Tools/AnalyzerRunner/AnalyzerRunner.csproj [hunk 7]
```diff
@@ -8,9 +8,35 @@
...
+  </Choose>
+
   <ItemGroup>
+    <Reference Include="System.ComponentModel.Composition" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="SQLitePCLRaw.bundle_green" Version="$(SQLitePCLRawbundle_greenVersion)" PrivateAssets="all" />
     <PackageReference Include="Microsoft.Build.Locator" Version="$(MicrosoftBuildLocatorVersion)" />
+    <PackageReference Include="SourceBrowser" Version="$(SourceBrowserVersion)" ExcludeAssets="build;compile" PrivateAssets="all" />
```
Comment (sharwell) [inline src/Tools/AnalyzerRunner/AnalyzerRunner.csproj]: 📝 I will file a follow-up issue to remove this once MSBuild Locator provides its own AppDomainManager.

File: src/Tools/AnalyzerRunner/IncrementalAnalyzerRunner.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+            workspace.Options = workspace.Options
+                .WithChangedOption(StorageOptions.SolutionSizeThreshold, _options.UsePersistentStorage ? 0 : int.MaxValue)
+                .WithChangedOption(RuntimeOptions.FullSolutionAnalysis, _options.FullSolutionAnalysis)
+                .WithChangedOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, LanguageNames.CSharp, _options.FullSolutionAnalysis)
+                .WithChangedOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, LanguageNames.VisualBasic, _options.FullSolutionAnalysis);
+
+            if (!string.IsNullOrEmpty(_options.ProfileRoot))
+            {
+                ProfileOptimization.StartProfile(nameof(IIncrementalAnalyzer));
```
Comment (genlu) [inline src/Tools/AnalyzerRunner/IncrementalAnalyzerRunner.cs]: What does `ProfileOptimization.StartProfile` do?

Comment (sharwell) [reply src/Tools/AnalyzerRunner/IncrementalAnalyzerRunner.cs]: Enables multi-core JIT to help reduce the impact of JIT on repeated measurements.

File: eng/targets/Settings.props (hunk 5 above)
Comment (JoeRobich) [reply eng/targets/Settings.props]: What does this change actually do? Was `*none*` a magic value before?

Comment (sharwell) [reply eng/targets/Settings.props]: There is no change in the behavior. MSBuild uses `Exists($(CodeAnalysisTargets))` for a condition. `*none*` does not exist, but unlike the new value, also contains `*` characters that are not valid. This change avoids a bunch of first chance exceptions that were making it hard for me to figure out why AnalyzerRunner was failing to load Roslyn.sln.

File: src/Tools/AnalyzerRunner/AnalyzerRunner.csproj (hunk 7 above)
Comment (sharwell) [reply src/Tools/AnalyzerRunner/AnalyzerRunner.csproj]: ➡️ Filed https://github.com/dotnet/roslyn/issues/36165

File: src/Tools/AnalyzerRunner/AnalyzerRunner.csproj (hunk 6 above)
Comment (tmat) [reply src/Tools/AnalyzerRunner/AnalyzerRunner.csproj]: Specifying `<PlatformTarget>AnyCPU</PlatformTarget>` should suffice.


---
In reply to: [289938504](https://github.com/dotnet/roslyn/pull/36113#discussion_r289938504) [](ancestors = 289938504)

Comment (sharwell) [reply src/Tools/AnalyzerRunner/AnalyzerRunner.csproj]: ➡️ Submitted #36192 to simplify this
